api/src/main/java/org/apache/iceberg/expressions/BoundPredicate.java

```diff
@@ -20,18 +20,35 @@
 package org.apache.iceberg.expressions;
 
 import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
 
-public abstract class BoundPredicate<T> extends Predicate<BoundReference<T>> {
-  protected BoundPredicate(Operation op, BoundReference<T> ref) {
-    super(op, ref);
+public abstract class BoundPredicate<T> extends Predicate<T, Bound<T>> implements Bound<Boolean> {
```

Comment:
`BoundPredicate` is `Expression` and `Bound` at the same time. Since `Bound` is `ValueExpression`, `BoundPredicate` is both `Expression` and `ValueExpression`. That seems to be in conflict as `ValueExpression` should evaluate to a value and `Expression` should evaluate to a boolean.

Reply:
Actually, why the original version of `Expression` is only about a boolean expression tree? Shouldn't that be called a predicate?

Reply:
Yeah, the original `Expression` interface is really a boolean expression. I didn't use "predicate" because boolean expressions include things like `true` and `and`.

But the point that we may want to clean this up is valid. I was thinking that it would make more sense if `ValueExpression` were called `Expression` - something that produces a value. Then `BooleanExpression` could be a `ValueExpression<Boolean>`. We'd need to remove the `ref` from the current `ValueExpression` and this would be a breaking API change.

Right now, `Bound` and `Unbound` extend only `ValueExpression` because `ValueExpression` has a reference. But `Bound` and `Unbound` could carry the reference instead.

Reply:
I think only `Bound` extends `ValueExpression` right now. That's where my original confusion  around `Bound` vs `Unbound` comes from.

Reply:
Bound probably doesn't need to extend ValueExpression.

Reply:
Okay, I cleaned this up quite a bit. I renamed `ValueExpression` to `Term`. That makes more sense because these are really one operand in a predicate, not a different kind of expression.

I also updated `Bound` and `Unbound` so they are independent, but I had to add `UnboundTerm` and `BoundTerm`. Overall, I think that's fine.

The structure is quite a bit cleaner now.